Multithreading

gbataille opened the pull request:
Hey Even,

So this is not to be merged but I wanted to share with you the work in progress to get some feedback.
So to implement multithreading for the image tiling, I'm starting by moving a lot of code outside of the Gdal2Tiles object, that is anyway more a global context than an object with clear intent (everything ends up in self, even some function local variables!). I don't have too but it's going to be cleaner AND will mean I don't have to worry about serializing a complex object like Gdal2Tiles.

2 issues with the current course of actions:
* Change in the output. In verbose output, you would get information about a tile, then generation of a tile, loop to the next tile. The new code, whether you would use 1 or more thread would do get info on all tiles, then generate all tiles.
Similarly, the progress indicator that exist today can be kept but at the cost of having a shared object in the middle (well I don't think it would cost too much, so I might do it).
I don't think that matters much. WDYT?
* It involves quite a bit of refactoring. While I ensure that test passes (obviously) and am testing manually the changes against some of my known datasets, I clearly don't know (and test) all the paths that exist in the code. Seeing the type of refactoring I'm doing for the moment, I figure that if I break something, it will be for all code paths, not just one, but who knows.
Because I don't understand (yet) all the things that the script can do, and because it would require too much energy/strength of will to first add the many tests that are missing, I decided to go like that, adding test only on the new things I'll create. Added to that, it's hard to review such code changes.

This strategy obviously has some risk so I wanted to check with you before I go further that you are ok with it

rouault commented:
> Change in the output. In verbose output, you would get information about a tile, then generation of a tile, loop to the next tile. The new code, whether you would use 1 or more thread would do get info on all tiles, then generate all tiles.
> Similarly, the progress indicator that exist today can be kept but at the cost of having a shared object in the middle (well I don't think it would cost too much, so I might do it).
> I don't think that matters much.

For verbose mode, no strong opinion. Might be a bit messy in multi-thread mode, but if that's people want, that's OK I guess. But a global progress is I think "required".

> It involves quite a bit of refactoring. While I ensure that test passes (obviously) and am testing manually the changes against some of my known datasets, I clearly don't know (and test) all the paths that exist in the code.

You can always call for testers... People should be earger to test and see all their cores be busy

rouault commented on gdal/swig/python/scripts/gdal2tiles.py:
```diff
@@ -956,6 +1235,8 @@ def open_input(self):
         if not self.out_ds:
             self.out_ds = self.in_ds
 
+        self.out_ds.GetDriver().CreateCopy("gba.vrt", self.out_ds)
```
Using a temporary file / unique filename would be better, otherwise there's a risk of collision if launching 2 gdal2tiles process from the same directory. For a single process approach, you could use a /vsimem/temporary_name.vrt to avoid hitting the disk. But if the file needs to be created once and shared by several processes (no idea here of what the code is doing !), then yes a on-disk file will be needed.

gbataille commented:
this is temporary. I definitely don't intend to leave my trigram in there :)

On Wed, Mar 8, 2017 at 1:47 PM, Even Rouault <notifications@github.com>
wrote:

> *@rouault* commented on this pull request.
> ------------------------------
>
> In gdal/swig/python/scripts/gdal2tiles.py
> <https://github.com/OSGeo/gdal/pull/201#discussion_r104909274>:
>
> > @@ -956,6 +1235,8 @@ def open_input(self):
>          if not self.out_ds:
>              self.out_ds = self.in_ds
>
> +        self.out_ds.GetDriver().CreateCopy("gba.vrt", self.out_ds)
>
> Using a temporary file / unique filename would be better, otherwise
> there's a risk of collision if launching 2 gdal2tiles process from the same
> directory. For a single process approach, you could use a
> /vsimem/temporary_name.vrt to avoid hitting the disk. But if the file needs
> to be created once and shared by several processes (no idea here of what
> the code is doing !), then yes a on-disk file will be needed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/201#pullrequestreview-25764046>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76V38jB7hEIs3lirCyhJ92GMak3Csks5rjqN-gaJpZM4MWXQ5>
> .
>

gbataille commented:
Ok so I moved along quite a bit.
I believe everything is functional but I want to do more test.

You'll see that all the blackbox test are still passing (I hope, they do run locally).

I also added a bunch of unit test, but I have not plugged them yet into travis.

The code is not 100% ready I would say, but it can serve as a basis, since reading the diff from the original file my be tricky now. There are a bunch of TODO with my name that I want to review before actually going forward.

You'll notice that

- I have copied the old script under gdal2tiles_old.py. I think this rewrite is so profound that we really need to have this safeguard for a couple of versions
- If you chose to run with a single process, I'm shortcutting all the `multiprocessing` code which means you should get similar behavior/performance as before the rewrite
- One of my big goal is to get rid of the big global GDAL2Tiles context object that brings nothing but complexity. Most of the features that I already extracted, I was able to test in python unittest

Waiting for your comments if you have time to do a first pass.

Cheers

rouault commented on gdal/swig/python/scripts/gdal2tiles.py:
```diff
@@ -462,65 +467,908 @@ def tilefilename(self, x, y, z):
+
+
+def options_post_processing(options, input_file, output_folder):
+    # Workaround for old versions of GDAL
```
I think you can remove such workarounds for old GDAL versions. gdal2tiles is released with a libgdal, so even requiring features only present in trunk would be OK

rouault commented on gdal/swig/python/scripts/gdal2tiles.py:
```diff
@@ -462,65 +467,908 @@ def tilefilename(self, x, y, z):
+    if details:
+        sys.stderr.write("\n\n%s\n" % details)
+
+    sys.exit(2)
```
it would be good to avoid sys.exit(), so that gdal2tiles can still be used as a library-like function

gbataille replied:
I see your point. This is however the existing behavior, albeit hidden behind calls to ArgumentParser.error which does just that.

I'll have a quick look to see how hard it would be to control those "exit conditions" in a different manner.

rouault commented on gdal/swig/python/scripts/gdal2tiles.py:
```diff
@@ -768,234 +1451,120 @@ def open_input(self):
+                self.warped_input_dataset = reproject_dataset(
+                    self.input_dataset, in_srs, self.out_srs)
+
+                if in_nodata:
```
I'm wondering if you couldn't avoid all those direct editing of the XML of the warped VRT by using gdal.Warp('out.vrt', in_dataset, options = '-of VRT other_options_here'), the librarified version of gdalwarp.

rouault replied:
Note: if you don't need on-disk storage of the VRT (which would be good, but for multi-threading requires to do it in each thread), you can even do gdal.Warp('/vsimem/' + some_unique_name_for_the_process, ...)

gbataille replied:
Completely agree. but let's leave that for potential later improvement if that's ok. There are a lot of things that can/should be done, but here I'd like to get the multithreading through.

gbataille commented:
Even, how do you propose we make this move forward?

I have rebased on master to re-trigger the tests, although I can't see that they had failed due to something I changed.

rouault commented:
There are still things to fix :
- gdal/swig/python/scripts/tests/gdal2tiles/test_add_gdal_warp_options_to_string.py:2: 'unittest.mock' imported but unused
- gdal/swig/python/scripts/tests/gdal2tiles/test_add_alpha_band_to_string_vrt.py:2: 'unittest.mock' imported but unused
- Windows tests fail with
``
line 2881, in single_threaded_tiling
    os.unlink(conf.src_file)
WindowsError: [Error 32] The process cannot access the file because it is being used by another process: 
```

gbataille commented:
ok, the windows build is good now!

rouault commented:
There are remaining test failures on precise_32bit and ubuntu_1604 targets:
```
Script: pyscripts/test_gdal2tiles.py
  TEST: test_python2_handles_utf8_by_default ... fail
    line 263: Should be handling filenames with utf8 characters in this context
  TEST: test_python2_gives_warning_if_bad_lc_ctype_and_non_ascii_chars ... fail
    line 263: Should be handling filenames with utf8 characters in this context
  TEST: test_python2_does_not_give_warning_if_bad_lc_ctype_and_non_ascii_chars_in_folder ... fail
    line 263: Should be handling filenames with utf8 characters in this context
```

gbataille commented:
Yup I saw yesterday evening. Not sure why this would be platform dependant.
I'll have a look

On Mon, 28 Aug 2017 at 22:11, Even Rouault <notifications@github.com> wrote:

> There are remaining test failures on precise_32bit and ubuntu_1604 targets:
>
> Script: pyscripts/test_gdal2tiles.py
>   TEST: test_python2_handles_utf8_by_default ... fail
>     line 263: Should be handling filenames with utf8 characters in this context
>   TEST: test_python2_gives_warning_if_bad_lc_ctype_and_non_ascii_chars ... fail
>     line 263: Should be handling filenames with utf8 characters in this context
>   TEST: test_python2_does_not_give_warning_if_bad_lc_ctype_and_non_ascii_chars_in_folder ... fail
>     line 263: Should be handling filenames with utf8 characters in this context
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/201#issuecomment-325466158>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76Swr5W_IfgFE6GhQMhdJt9CpgV4jks5scx7kgaJpZM4MWXQ5>
> .
>
-- 
Greg

rouault commented:
One common point of precise_32bit and ubuntu_1604 targets is that they run with "export LC_ALL=en_US". I'm not sure if/why it is needed (I think it was needed for installing packages in the chroot, but perhas not for running test). Cf ci/travis/ubuntu_1604/script.sh and ci/travis/precise_32bit/script.sh. You could perhaps try removing it

gbataille commented:
oh.... that might have an impact.
Thanks for pointing it out

On Tue, Aug 29, 2017 at 12:46 PM, Even Rouault <notifications@github.com>
wrote:

> One common point of precise_32bit and ubuntu_1604 targets is that they run
> with "export LC_ALL=en_US". I'm not sure if/why it is needed (I think it
> was needed for installing packages in the chroot, but perhas not for
> running test). Cf ci/travis/ubuntu_1604/script.sh and
> ci/travis/precise_32bit/script.sh. You could perhaps try removing it
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/201#issuecomment-325626693>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76VTkvXqNzGLvyxslJ1UTSJzkVH3uks5sc-wVgaJpZM4MWXQ5>
> .
>

gbataille commented:
still haven't had the time to dig into that. However, I wanted to mention that I have installed this version of gdal2tiles on our prod.

We have basically one use case: from geotiff (4 bands, or 1 band) to mercator. But we are doing ~500 geotiff tiling a day and it seems that it works as expected.

rouault commented:
@gbataille Can you rebase on trunk ? I've done changes in ci/travis/* files to generate the locale files, so the locale / UTF8 related errors should be solved now

rouault commented:
I've restart the Mac job, but I'm seeing the same timeouts on trunk too, so this is unrelated to your work and we're probably all good for merging !

rouault commented:
ok, I've merged it. Travis has currently issues with Mac workers. Thanks for your work !

rouault commented:
Ah, and it would make sense that you announce it on the gdal-dev list

rouault commented:
@gbataille 

I added myself in test_gdal2tiles_py_zoom_option as '--processes=2' switch as multithreading didn't seem to be actually tested. It worked fine on some Travis workers, but failed on a fair number (see https://travis-ci.org/OSGeo/gdal/builds/278755009 ) with
```
  TEST: test_gdal2tiles_py_zoom_option ... fail (blowup)
Traceback (most recent call last):
  File "pymod/gdaltest_python2.py", line 42, in run_func
    result = func()
  File "./test_gdal2tiles.py", line 84, in test_gdal2tiles_py_zoom_option
    '-q -z 0-1 --processes=2 ../gdrivers/data/small_world.tif tmp/out_gdal2tiles_smallworld')
  File "../pymod/test_py_scripts.py", line 80, in run_py_script
    return run_py_script_as_py_module(script_path, script_name, concatenated_argv)
  File "../pymod/test_py_scripts.py", line 141, in run_py_script_as_py_module
    exec(script_name + '.main()')
  File "<string>", line 1, in <module>
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal2tiles.py", line 2939, in main
    multi_threaded_tiling(input_file, output_folder, options)
  File "/home/travis/build/OSGeo/gdal/autotest/pyscripts/../../gdal/swig/python/scripts/gdal2tiles.py", line 2906, in multi_threaded_tiling
    queue = manager.Queue()
  File "/usr/lib/python2.7/multiprocessing/managers.py", line 667, in temp
    token, exp = self._create(typeid, *args, **kwds)
  File "/usr/lib/python2.7/multiprocessing/managers.py", line 565, in _create
    conn = self._Client(self._address, authkey=self._authkey)
  File "/usr/lib/python2.7/multiprocessing/connection.py", line 175, in Client
    answer_challenge(c, authkey)
  File "/usr/lib/python2.7/multiprocessing/connection.py", line 413, in answer_challenge
    message = connection.recv_bytes(256)         # reject large message
IOError: [Errno 11] Resource temporarily unavailable
```

Any idea why ?

rouault commented:
Different failure mode on Windows when testing --processes in autotest
```
Running tests from pyscripts/test_gdal2tiles.py
  TEST: test_gdal2tiles_py_simple ... success
  TEST: test_gdal2tiles_py_zoom_option ... Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "c:\Python27-x64\lib\multiprocessing\forking.py", line 380, in main
    prepare(preparation_data)
  File "c:\Python27-x64\lib\multiprocessing\forking.py", line 503, in prepare
    file, path_name, etc = imp.find_module(main_name, dirs)
ImportError: No module named run_all
```
(the reference to run_all must be the autotest/run_all.py script)

gbataille commented:
hum, thanks for adding those tests. I should have done that indeed. no clue
what's happening. I'll try to have a look this week end if I can.

On Fri, Sep 22, 2017 at 11:48 PM, Even Rouault <notifications@github.com>
wrote:

> Different failure mode on Windows when testing --processes in autotest
>
> Running tests from pyscripts/test_gdal2tiles.py
>   TEST: test_gdal2tiles_py_simple ... success
>   TEST: test_gdal2tiles_py_zoom_option ... Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>   File "c:\Python27-x64\lib\multiprocessing\forking.py", line 380, in main
>     prepare(preparation_data)
>   File "c:\Python27-x64\lib\multiprocessing\forking.py", line 503, in prepare
>     file, path_name, etc = imp.find_module(main_name, dirs)
> ImportError: No module named run_all
>
> (the reference to run_all must be the autotest/run_all.py script)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/201#issuecomment-331568465>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76cPmPAmxnJegRPARN5BLeqzvQknnks5slCshgaJpZM4MWXQ5>
> .
>

gbataille commented:
Hey Even,

Have you seen that anywhere else than travis? I'm trying to reproduce it
but with no luck.
I tried to download and run on an ubuntu 12 VM but I don't get the issue.
and running the travis install.sh scripts seems to be very painful and
necessitates some basic box config that I don't know.

However the issue is not completely unexpected as multiprocessing has known
to be flaky (especially on old config) and that's why I went through the
trouble of having a different code when we use a single process, to not use
any multiprocessing artifacts at all.

I'll try to continue to reproduce it, but I don't really know what to do
next.

On Fri, Sep 22, 2017 at 11:59 PM, Grégory Bataille <gbataille.dev@gmail.com>
wrote:

> hum, thanks for adding those tests. I should have done that indeed. no
> clue what's happening. I'll try to have a look this week end if I can.
>
> On Fri, Sep 22, 2017 at 11:48 PM, Even Rouault <notifications@github.com>
> wrote:
>
>> Different failure mode on Windows when testing --processes in autotest
>>
>> Running tests from pyscripts/test_gdal2tiles.py
>>   TEST: test_gdal2tiles_py_simple ... success
>>   TEST: test_gdal2tiles_py_zoom_option ... Traceback (most recent call last):
>>   File "<string>", line 1, in <module>
>>   File "c:\Python27-x64\lib\multiprocessing\forking.py", line 380, in main
>>     prepare(preparation_data)
>>   File "c:\Python27-x64\lib\multiprocessing\forking.py", line 503, in prepare
>>     file, path_name, etc = imp.find_module(main_name, dirs)
>> ImportError: No module named run_all
>>
>> (the reference to run_all must be the autotest/run_all.py script)
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/OSGeo/gdal/pull/201#issuecomment-331568465>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ABD76cPmPAmxnJegRPARN5BLeqzvQknnks5slCshgaJpZM4MWXQ5>
>> .
>>
>
>

rouault commented:
For the Ubuntu 16.04 failure
```
    sl = self._semlock = _multiprocessing.SemLock(kind, value, maxvalue)
OSError: [Errno 38] Function not implemented
```
https://stackoverflow.com/questions/6033599/oserror-38-errno-38-with-multiprocessing points to /dev/shm not being accessible, which is probably because https://github.com/OSGeo/gdal/blob/trunk/gdal/ci/travis/ubuntu_1604/before_install.sh#L11 doesn't mount it. So probably that an extra mount for host /dev/shm to guest could fix that
(you can actually run quite easily this config by running manually the scripts before_install.sh, install.sh, script.sh if your OS is Debian/Ubuntu based since it is basically doing a chroot install. do that in a dedicated checkout of the GDAL repo since that will create files with root permission)

The 4 other failed targets all use a Travis precise VM and all fail with
```
File "/usr/lib/python2.7/multiprocessing/connection.py", line 413, in answer_challenge
    message = connection.recv_bytes(256)         # reject large message
IOError: [Errno 11] Resource temporarily unavailable
```
I found https://bugs.python.org/issue6056 which seems to point to using socket.setdefaulttimeout(), which some GDAL autotests do use by calling https://github.com/OSGeo/gdal/blob/trunk/autotest/pymod/gdaltest_python2.py#L67
Perhaps calling socket.setdefaulttimeout(None) before the test would fix it. I cannot reproduce quickly since my host is a Ubuntu 16
I'm going to try in a branch

rouault commented:
I've finally fixed the issue on all platforms, including Windows, with https://trac.osgeo.org/gdal/changeset/40229 , by running the test as an external process, instead from the parent process

Do not hesitate to advertize your work on gdal-dev. I'm pretty sure a lot of people will be interested.

gbataille commented:
wow, thanks, had not found those articles.

I had not been able to run the before_install.sh and co on my VMs, not sure
why. Maybe not starting from a compatible images.
Thanks for the help, I would have needed a very long time to make all that
work.

Cheers

On Wed, Sep 27, 2017 at 1:46 PM, Even Rouault <notifications@github.com>
wrote:

> I've finally fixed the issue on all platforms, including Windows, with
> https://trac.osgeo.org/gdal/changeset/40229 , by running the test as an
> external process, instead from the parent process
>
> Do not hesitate to advertize your work on gdal-dev. I'm pretty sure a lot
> of people will be interested.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/201#issuecomment-332495414>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76dzvO4tkCFi3DkcM8zoEwtoIk0wAks5smjWZgaJpZM4MWXQ5>
> .
>

jo-chemla commented:
Hi Even & Gregory, 
I would love to try out the multi-process version of gdal2tiles - I've upgraded my machine, but the new one get an error message when using gdal2tiles, `Incompatible libpng version in application and library`, so before fixing it I would like to try the parallel version on the older computer. 
What is the recommended way to try it? I cannot find the instructions to build from source on windows, and most gudies are outdated, and it seems existing python wheels do not incorporate the multi-process enhancements. 
EDIT: I've found it however in the OsGeo4W64 installer, it seems to be this multiprocess version in the dev gdal package. But right now I have the same libpng error, I will try out on the older computer!

gbataille commented:
Hi

Are you able to install any *2.x* version of GDAL and the corresponding
python bindings?
If you can, then you can grab the gdal2tiles.py file that is in the github
repository
https://raw.githubusercontent.com/OSGeo/gdal/trunk/gdal/swig/python/scripts/gdal2tiles.py
and just use it.
It's a standalone script that should be able with any install of gdal 2.x I
think

On Thu, Oct 5, 2017 at 5:53 PM, jo-chemla <notifications@github.com> wrote:

> Hi Even & Gregory,
> I would love to try out the multi-process version of gdal2tiles - I've
> upgraded my machine, but the new one get an error message when using
> gdal2tiles, Incompatible libpng version in application and library, so
> before fixing it I would like to try the parallel version on the older
> computer.
> What is the recommended way to try it? I cannot find the instructions to
> build from source on windows, and most gudies are outdated, and it seems
> existing python wheels do not incorporate the multi-process enhancements.
> EDIT: I've found it however in the OsGeo4W64 installer, it seems to be
> this multiprocess version in the dev gdal package. But right now I have the
> same libpng error, I will try out on the older computer!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/201#issuecomment-334509234>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76Y6CcVvnKQFpE4Jhev9ie87kw8thks5spPuBgaJpZM4MWXQ5>
> .
>

jo-chemla commented:
Thanks, seems to be working, all cores hitting 100% and processing time decreased a lot! That's a very welcome addition. 

And you are right, with basic GDAL 2.* on two different computers, I cannot get gdal2tiles to work, getting a version incompatibility with libpng listed below, and nearly no one has the same issue on Windows. The error happens when the code tries to call `self.out_drv.CreateCopy(tilefilename, dstile, strict=0)`, I need to investigate further. Thanks again!

```
Generating Base Tiles:
Warning 1: libpng: Application was compiled with png.h from libpng-1.6.32
Warning 1: libpng: Application  is  running with png.c from libpng-1.2.56
ERROR 1: libpng: Incompatible libpng version in application and library
```

rouault commented:
> Is there something I'm missing?

There was a performance issue I've fixed per 0b89f73c7fdbd16e281dfb50725e14e3e68d07fe. Will be available in GDAL 2.3.2 (or you can just grab the updated script from https://github.com/OSGeo/gdal/blob/master/gdal/swig/python/scripts/gdal2tiles.py

rogg commented:
Yeah, using 2.3.1 with the current master version of gdal2tiles is where I got those benchmarks from. I built gdal 2.3.2rc1 and now it beats the old version. Thanks!

gbataille commented:
Thanks Even,

Must say I never followed up on the rewrite I did, but I did not think I
had degraded the performance, not that much at least. My specific dataset
seemed fine, but maybe I was in some special cases.
Sorry

On Sun, Sep 23, 2018 at 4:06 PM rogg <notifications@github.com> wrote:

> Yeah, using 2.3.1 with the current master version of gdal2tiles is where I
> got those benchmarks from. I built gdal 2.3.2rc1 and now it beats the old
> version. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OSGeo/gdal/pull/201#issuecomment-423819066>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABD76WIhNc083JYJ_Icuubl-aRkEeRyiks5ud5VWgaJpZM4MWXQ5>
> .
>

cr8tpro commented:
Hi @gbataille. Thanks for your great work.
I just have a question with --processes parameter in gdal2tiles.py
I am running AWS EC2 instance for imagery tiling using gdal2tiles.py and it has 96vCpus and 384 cores.
In this case, what number should I choose for parameter of --processes?
I want to use EC2 instance performance as much as I can.

Regards
Bojan Stojanovic

gbataille commented:
Hey @cr8tpro ,

It's not really possible to give a definite answer.
the `--processes` uses a divide and conquer approach. With 96 vCPUs (wow you do use a big machine), you can theoretically use ~96 processes without them competing for resources.
But the problem is that splitting the image in 96 chunks has an overhead. And reading the same file from 96 different places too.

I must say that those questions can only be answered by experimentation, as it depends on the source dataset and on the properties of the hardware you use (not only the number of CPUs, but the disk bandwidth, ...)
I have never used more than 4 myself

I would apply some sampling and then dichotomy. seeing performance with say 4, 10, 25, 50, 75, 96
I guess the higher ones will either not be better, or even be worse.
And then you can refine the sweetspot a bit.

But I fear there is no definite answer and a larger image might benefit from a few more processes while a smaller image will just suffer from being split too far

Sorry I don't have a clean answer
